magic_the_gathering/phases/combat_declare_attackers.py

```diff
@@ -1,13 +1,28 @@
+from typing import List
+
+from magic_the_gathering.actions.base import Action
+from magic_the_gathering.actions.declare_attacker import DeclareAttackerAction
 from magic_the_gathering.game_state import GameState
-from magic_the_gathering.phases.base import Phase
+from magic_the_gathering.phases.players_get_priority import PhaseWherePlayersGetPriority
+
 
+class CombatDeclareAttackersPhase(PhaseWherePlayersGetPriority):
+    @staticmethod
+    def list_possible_actions(game_state: GameState) -> List[Action]:
+        return DeclareAttackerAction.list_possible_actions(game_state=game_state)
```

Comment:
If I understood the code well, there is an issue here: here this list is a list of `this creature attacks this player` like
- creature a attacks player B
- creature a attacks player C
- creature b attacks player B
- creature b attacks player C
And then the random IA will choose one of these.
But isn't what we want to choose at random a set of "this creature attacks this player" rather than a single one?
Like, option 1 could be "a attacks B and b attacks C"
Also, we need to support options where a creature does not attack, like the set "a attacks B"

Reply:
Yes you are right, the players needs to be able to declare several attackers. The way I chose to do it (because I found it easier to implement although I do understand that it is not clear without explanations) is to use a while loop in the run method:
```python
def run(self, game_state: GameState) -> GameState:
        while True:
            possible_actions = CombatDeclareAttackersPhase.list_possible_actions(game_state)
            current_player = game_state.current_player
            action = current_player.choose_action(
                game_state=game_state,
                possible_actions=possible_actions,
            )
            if action is None:
                break
            game_state = action.execute(game_state)
        return game_state
```
Here is the process:
- Current player gets to choose one attack action among the possible actions
- If the chosen action is None, we consider the player stops declaring attackers (btw I realize that I did not include this in the implementation of the Player, so you are right it would be nice to have an action to mean "I don't want to declare more attackers")
- Otherwise, the chosen action is applied. You can see how I implement the declaration of the attackers in `magic_the_gathering/actions/declare_attacker.py`.
- Now we enter the loop again, we list the possible actions (which is different since the previous action affected the game state so that one card is already an attacker and thus cannot be chosen again)
- The current player gets to choose whether it declares one more attacker
- Etc...

For me it was simpler to implement it this way. The `DeclareAttackerAction` is more "atomic", we support the declaration of several attackers and it means that on the player side everything looks like choosing one action at a time (which is simpler than deciding a set of actions at once).

However, if you find issues with this implementation, we can discuss of course :slightly_smiling_face:

Reply:
I think I found how to handle the "I don't want to declare attackers anymore":
```python
@staticmethod
    def list_possible_actions(game_state: GameState) -> List[Action]:
        possible_actions = [None]
        possible_actions.extend(DeclareAttackerAction.list_possible_actions(game_state=game_state))
        return possible_actions
```

If there is one `None` element in the list, then the player can choose it and then it will be interpreted in the run method as I explained before :slightly_smiling_face:

Reply:
I added `None` to the possible actions in all the relevant phases. Of course we can change this if you think there is a better way :wink:

Reply:
Ah ok
One problem I can think of with this approach is that it seems that there is a need to check if "sub-battle" is legal, with a sub-battle being one attacker being blocked by one or more blockers.
Ex: `menace` makes a creature only be blockable by 2 creatures or more.
This makes it so some combination of blocks may be individually legal, but not legal when looked at together.
And menace is common.
Like an attacker with menace attacks, I have a single creature, I list the possible first blocks I could take, I block it with one creature, but then I can't block it with more, and my blocking combination is illegal.

Reply:
Ah, I see! I was not aware of that. I will create an issue for this, thank you :slightly_smiling_face: